providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,334 @@
+        """
+        # nvidia-smi ship with NVIDIA GPU display drivers on Linux
+        # https://developer.nvidia.com/nvidia-system-management-interface
+        # check prime-select to make sure system with nv driver.
+        # If there is no nv driver, prime offload is fine for other drivers.
+        ps = subprocess.run("whereis prime-select | cut -d ':' -f 2",
+                            shell=True,
+                            stdout=subprocess.PIPE,
+                            universal_newlines=True)
+        if 'prime-select' not in ps.stdout:
+            return PrimeOffloaderError.NO_ERROR
+
+        # prime offload could running on on-demand mode only
+        mode = subprocess.run("prime-select query",
+                              shell=True,
+                              stdout=subprocess.PIPE,
+                              universal_newlines=True)
+        if "on-demand" not in mode.stdout:
+            self.logger.info("System isn't on-demand mode")
+            return PrimeOffloaderError.NOT_SUPPORT_NV_PRIME
+
+        # prime offload couldn't running on nvlink active or inactive
+        # Therefore, only return empty string is supported environment.
+        nvlink = subprocess.run("nvidia-smi nvlink -s",
+                                shell=True,
+                                stdout=subprocess.PIPE,
+                                universal_newlines=True)
+        if len(nvlink.stdout) != 0:
+            if re.search('error', nvlink.stdout, re.IGNORECASE):
+                self.logger.info("nvidia driver error")
+                return PrimeOffloaderError.NV_DRIVER_ERROR
+            self.logger.info("NVLINK detected")
+            return PrimeOffloaderError.NOT_SUPPORT_NV_PRIME
+
+        return PrimeOffloaderError.NO_ERROR
+
+    def run_offload_cmd(self, cmd, pci_name, driver, timeout):
+        """
+        run offload command and check it runs on correct GPU
+
+        :param cmd: command that running under prime offload
+        :type cmd: str
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type pci_name: str
+
+        :param driver: GPU driver, such as i915, amdgpu, nvidia
+        :type driver: str
+        """
+        card_id = self.find_card_id(pci_name)
+        if card_id == PrimeOffloaderError.NO_CARD_ID:
+            self.logger.info("Couldn't find card id,"
+                             " please check your pci name")
+            return PrimeOffloaderError.NO_CARD_ID
+
+        card_name = self.find_card_name(pci_name)
+        if card_name is None:
+            self.logger.info("Couldn't find card name,"
+                             " please check your pci name")
+            return PrimeOffloaderError.NO_CARD_NAME
+
+        # run offload command in other process
+        dri_pci_name_format = re.sub("[:.]", "_", pci_name)
+
+        if timeout > 0:
+            tmp_cmd = "timeout {} {}".format(timeout, cmd)
+        else:
+            tmp_cmd = cmd
```

Comment:
No need to create another variable, just prepend it with timeout if needed.
```
if timeout > 0:
    cmd = "timeout {} {}".format(timeout, cmd)
```
or even better, when using a lexically structured command:
```
if timeout:
    cmd = ['timeout', timeout] + cmd
```
another benefit of structured cmds

Reply:
I have to use original cmd for checking this command is running.
```python
# use other thread to check offload is correctly or not
check_thread = threading.Thread(target=self.check_offload,
                                                      args=(cmd, card_id, card_name))
check_thread.start()
```